test_palindrome.py

```diff
@@ -0,0 +1,151 @@
+def test_palindromes_option_case_sensitive(whitespace_sensitive: bool, punctuation_sensitive: bool) -> None:
+    TestCaseCaseSensitive = namedtuple(
+        typename='TestCaseCaseSensitive',
+        field_names=["expectation", "data", "case_sensitive"]
+    )
+    test_cases = [
+        TestCaseCaseSensitive(expectation=False, data="Atak kata", case_sensitive=True),
+        TestCaseCaseSensitive(expectation=True, data="Atak kata", case_sensitive=False),
+    ]
+    for expectation, data, case_sensitive in test_cases:
+        result = is_palindrome(data, case_sensitive, whitespace_sensitive, punctuation_sensitive)
+        assert result == expectation, \
+            f"Expected '{expectation}', got '{result}' for '{data=}' and "\
+            f"options '{case_sensitive=}', '{whitespace_sensitive=}', '{punctuation_sensitive=}'!"
+
+
+@pytest.mark.parametrize("case_sensitive", [True, False])
+@pytest.mark.parametrize("punctuation_sensitive", [True, False])
+def test_palindromes_option_whitespace_sensitive(case_sensitive: bool, punctuation_sensitive: bool) -> None:
+    TestCaseWhitespaceSensitive = namedtuple(
+        typename='TestCaseWhitespaceSensitive',
+        field_names=["expectation", "data", "whitespace_sensitive"],
+    )
+    test_cases = [
+        TestCaseWhitespaceSensitive(expectation=False, data="moc owocom", whitespace_sensitive=True),
+        TestCaseWhitespaceSensitive(expectation=True, data="moc owocom", whitespace_sensitive=False),
+    ]
+    for expectation, data, whitespace_sensitive in test_cases:
+        result = is_palindrome(data, case_sensitive, whitespace_sensitive, punctuation_sensitive)
+        assert result == expectation, \
+            f"Expected '{expectation}', got '{result}' for '{data=}' and "\
+            f"options '{case_sensitive=}', '{whitespace_sensitive=}', '{punctuation_sensitive=}'!"
+
+
+@pytest.mark.parametrize("case_sensitive", [True, False])
+@pytest.mark.parametrize("whitespace_sensitive", [True, False])
+def test_palindromes_option_punctuation_sensitive(case_sensitive: bool, whitespace_sensitive: bool) -> None:
+    TestCasePunctuationSensitive = namedtuple(
+        typename='TestCasePunctuationSensitive',
+        field_names=["expectation", "data", "punctuation_sensitive"],
+    )
+    test_cases = [
+        TestCasePunctuationSensitive(expectation=False, data="atak kata.", punctuation_sensitive=True),
+        TestCasePunctuationSensitive(expectation=True, data="atak kata.", punctuation_sensitive=False),
+    ]
+    for expectation, data, punctuation_sensitive in test_cases:
+        result = is_palindrome(data, case_sensitive, whitespace_sensitive, punctuation_sensitive)
+        assert result == expectation, \
+            f"Expected '{expectation}', got '{result}' for '{data=}' and "\
+            f"options '{case_sensitive=}', '{whitespace_sensitive=}', '{punctuation_sensitive=}'!"
+
+
+def test_palindromes_all_options_mixed() -> None:
```

Comment:
Tak, myślę, że taki test jest celowy. Więc jeśli chciałbyś to zrobić za pomocą `pytest.mark.parametrize` to tak jak zrobiłes to poniżej, trzeba by przekazywać aż pięć różnych parametrów (ewentualnie cztery jeśli `data` się nie zmienia).

Reply:
Zauważ, że jeśli przekazujesz coś za pomocą `parametrize`, to pytest pokaże Ci informację o tym, jaka kombinacja "pada", więc nie jest konieczne pisanie aż takich rozbudowanych opisów błędów.